inc/class-wpseo-utils.php

```diff
@@ -483,35 +495,276 @@ public static function clear_transient_cache( $option ) {
-			}
+		// Refresh validator.
+		self::new_sitemap_cache_validator( $type );
+
+		if ( ! wp_using_ext_object_cache() ) {
+			// Clean up current cache from the database.
+			self::cleanup_sitemap_database_cache( $type, $old_validator );
+		}
+
+		// External object cache pushes old and unretrieved items out by itself so we don't have to do anything for that.
+	}
+
+	/**
+	 * Cleanup invalidated database cache
+	 *
+	 * @param null|string $type      The type of sitemap to clear cache for.
+	 * @param null|string $validator The validator to clear cache of.
+	 *
+	 * @return void
+	 */
+	private static function cleanup_sitemap_database_cache( $type = null, $validator = null ) {
+		global $wpdb;
+
+		if ( is_null( $type ) ) {
+			// Clear all cache if no type is provided.
+			$like = sprintf( '_transient_%s%%', self::$sitemap_cache_key_prefix );
 		}
 		else {
-			$query .= " option_name LIKE '_transient_wpseo_sitemap_%' OR option_name LIKE '_transient_timeout_wpseo_sitemap_%'";
+			if ( ! is_null( $validator ) ) {
+				// Clear all cache for provided type-validator.
+				$like = sprintf( '_transient_%%_%s', $validator );
+			}
+			else {
+				// Clear type cache for all type keys.
+				$like = sprintf( '_transient_%s%s_%%', self::$sitemap_cache_key_prefix, $type );
+			}
 		}
 
+		/**
+		 * Add slashes to the LIKE "_" single character wildcard.
+		 *
+		 * We can't use `esc_like` here because we need the % in the query.
+		 */
+		$where = sprintf( "option_name LIKE '%s'", addcslashes( $like, '_' ) );
+
+		$query = sprintf( 'DELETE FROM %s WHERE %s', $wpdb->options, $where );
 		$wpdb->query( $query );
 	}
 
 	/**
+	 * Get the sitemap cache key prefix
+	 *
+	 * @return string
+	 */
+	public static function get_sitemap_cache_key_prefix() {
+		return self::$sitemap_cache_key_prefix;
+	}
+
+	/**
+	 * Get the cache key for a certain type and page
+	 *
+	 * Example key format for type "post", page 1: wpseo_sitemap_post_1:akfw3e_23azBa
+	 *
+	 * @param null|string $type The type to get the key for. Null or self::SITEMAP_INDEX_TYPE for index cache.
+	 * @param int         $page The page of cache to get the key for.
+	 *
+	 * @return string The key where the cache is stored on.
+	 */
+	public static function get_sitemap_cache_key( $type = null, $page = 1 ) {
+		// Using SITEMAP_INDEX_TYPE for sitemap index cache.
+		$type = is_null( $type ) ? self::SITEMAP_INDEX_TYPE : $type;
+
+		$global_cache_validator = self::get_sitemap_cache_validator();
+		$type_cache_validator   = self::get_sitemap_cache_validator( $type );
+
+		$prefix = self::$sitemap_cache_key_prefix;
+		$postfix = sprintf( '_%d:%s_%s', $page, $global_cache_validator, $type_cache_validator );
+
+		$type = self::get_safe_sitemap_cache_type( $type, $prefix, $postfix );
+
+		// Build key.
+		$full_key = $prefix . $type . $postfix;
```

Comment:
Why not return this immediately instead of creating a new variable and returning that?